### PR TITLE
Collector Service (Schema based Ingest)

### PR DESCRIPTION
**Pull request by drtyh2o**
API that can accept any data stream, validate messages against a JSON schema, archive them, and then stream them via Kafka where they can be consumed in a type-safe manner.

### CODE REVIEW

**Comment by drtyh2o** [comment]
@karcass @avatarneil This is at a good spot now. API endpoint that accepts messages, validates them against a JSON schema object and if valid, persists them, and streams them over Kafka.

File: packages/mds-collector-api/handlers/write-schema-messages.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  next: NextFunction
+) => {
+  try {
+    const { schema_id } = req.params
+    if (!isNonEmptyArray(req.body)) {
+      throw new ValidationError('Request must contain a non-empty array of messages')
+    }
+    // eslint-reason checkAccess middleware has previously verified that local.claims.provider_id is a UUID
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const producer_id = res.locals.claims!.provider_id!
```
Comment (avatarneil) [inline packages/mds-collector-api/handlers/write-schema-messages.ts:46]: I worry that `producer_id` could get a bit confusing when also mixing in Kafka (where there are already Producers)... Perhaps `principal_id` could be used again here?

File: packages/mds-collector-backend/@types/index.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+import { RpcRoute, RpcServiceDefinition } from '@mds-core/mds-rpc-common'
+import { DomainModelCreate } from '@mds-core/mds-repository'
+import { NonEmptyArray, Timestamp, UUID } from '@mds-core/mds-types'
+
+export interface CollectorMessageDomainModel {
+  schema_id: string
+  producer_id: UUID
+  message: {}
```
Comment (avatarneil) [inline packages/mds-collector-backend/@types/index.ts:24]: I think you probably want `object` as the type here, not `{}`. More restrictive

File: packages/mds-collector-backend/service/provider.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  },
+  writeSchemaMessages: async (schema_id, producer_id, messages) => {
+    try {
+      const [validate, producer] = await Promise.all([getValidateFunction(schema_id), getStreamProducer(schema_id)])
+
+      const invalid = messages.reduce<{ position: number; errors: Partial<ErrorObject>[] }[]>(
+        (errors, message, position) => {
+          // eslint-reason validate function has previously verified that errors is non-null
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          return validate(message) ? errors : errors.concat({ position, errors: validate.errors! })
```
Comment (avatarneil) [inline packages/mds-collector-backend/service/provider.ts:118]: I recommend we use the AJV `async` validators here. The synchronous API is honestly pretty abysmal (as demonstrated by the need to do a non-null assertion here)

File: packages/mds-collector-backend/service/provider.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+          type: 'ValidationError',
+          message: `Invalid ${pluralize(invalid.length, 'message', 'messages')} for schema ${schema_id}`,
+          details: { invalid }
+        })
+      }
+
+      // Write to Postgres
+      const result = await CollectorRepository.insertCollectorMessages(
+        messages.map(message => ({ schema_id, producer_id, message })),
+        // Write to Kafka prior to committiing transaction
```
Comment (avatarneil) [inline packages/mds-collector-backend/service/provider.ts:134]:
```suggestion
        // Write to Kafka prior to committing transaction
```

File: packages/mds-collector-backend/repository/index.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+              manager
+                .getRepository(CollectorMessageEntity)
+                .createQueryBuilder()
+                .insert()
+                .values(messages.map(CollectorMessageDomainToEntityCreate.mapper()))
+                .returning('*')
+                .execute()
+            )
+          )
+          await beforeCommit()
```
Comment (avatarneil) [inline packages/mds-collector-backend/repository/index.ts:48]: I think that having this closure is probably a little excessive... Is there really much value in creating the transaction, then running a closure (as used so far, to write to Kafka), then committing the transaction? I don't see much value that is added as opposed to just running the Kafka writes after properly persisting the data in PG (or beforehand, if you don't care about it being persisted in PG); and the usage looks pretty weird (especially given that JS doesn't have named parameters), it seems as if it's a CB function but it.. isn't really?

File: packages/mds-collector-api/handlers/write-schema-messages.ts (hunk 1 above)
Comment (drtyh2o) [reply packages/mds-collector-api/handlers/write-schema-messages.ts:46]: The idea was `producer_id` identified the source of the data. Perhaps we should just stick with `provider_id` as in an entity that's providing data to MDS.

File: packages/mds-collector-backend/@types/index.ts (hunk 2 above)
Comment (drtyh2o) [reply packages/mds-collector-backend/@types/index.ts:24]: The idea is to allow anything that can be described by a schema, but I can change to `object` if we don't think we want to ingest strings or numbers alone.

File: packages/mds-collector-backend/repository/index.ts (hunk 5 above)
Comment (drtyh2o) [reply packages/mds-collector-backend/repository/index.ts:48]: The idea was that the PG and Kafka writes both succeed so the Kafka writes are done within the transaction. I could change the second parameter to `options: Partial<{ beforeCommit: () => Promise<void> }> = {}`

File: packages/mds-collector-backend/service/provider.ts (hunk 3 above)
Comment (drtyh2o) [reply packages/mds-collector-backend/service/provider.ts:118]: The `async` API is kinda weird too. I'll look into abstracting the Ajv API in the SchemaValidator to provide a better interface. We can avoid the non-null assertion without resorting to async validation.

**Review by nessur — APPROVED** [review]

**Review by avatarneil — APPROVED** [review]

**Review by nessur — APPROVED** [review]

**Pull request merged**